ksqldb-api-client/src/main/java/io/confluent/ksql/api/client/impl/ClientImpl.java

```diff
@@ -0,0 +1,207 @@
+  public CompletableFuture<List<Row>> executeQuery(final String sql) {
+    return executeQuery(sql, Collections.emptyMap());
+  }
+
+  @Override
+  public CompletableFuture<List<Row>> executeQuery(
+      final String sql,
+      final Map<String, Object> properties
+  ) {
+    return makeQueryRequest(sql, properties, PullQueryResponseHandler::new);
+  }
+
+  @Override
+  public CompletableFuture<Void> insertInto(
+      final String streamName, final Map<String, Object> row) {
+    return null; // not yet implemented
+  }
+
+  @Override
+  public Publisher<InsertAck> streamInserts(
+      String streamName, Publisher<List<Object>> insertsPublisher) {
+    return null; // not yet implemented
+  }
+
+  @Override
+  public void close() {
+    httpClient.close();
+    if (ownedVertx) {
+      vertx.close();
+    }
+  }
+
+  @FunctionalInterface
+  private interface ResponseHandlerSupplier<T> {
+    QueryResponseHandler<T> get(Context ctx, RecordParser recordParser, CompletableFuture<T> cf);
+  }
+
+  private <T> CompletableFuture<T> makeQueryRequest(
+      final String sql,
+      final Map<String, Object> properties,
+      final ResponseHandlerSupplier<T> responseHandlerSupplier
+  ) {
+
+    final JsonObject requestBody = new JsonObject().put("sql", sql).put("properties", properties);
+
+    final CompletableFuture<T> cf = new CompletableFuture<>();
+
+    HttpClientRequest request = httpClient.request(HttpMethod.POST,
+        serverSocketAddress, clientOptions.getPort(), clientOptions.getHost(),
+        "/query-stream",
+        response -> handleResponse(response, cf, responseHandlerSupplier))
+        .exceptionHandler(e -> handleRequestException(e, cf));
+    if (clientOptions.isUseBasicAuth()) {
+      request = configureBasicAuth(request);
+    }
+    request.end(requestBody.toBuffer());
+
+    return cf;
+  }
+
+  private HttpClientRequest configureBasicAuth(final HttpClientRequest request) {
+    final String creds = clientOptions.getBasicAuthUsername()
+        + ":"
+        + clientOptions.getBasicAuthPassword();
+    final String base64creds =
+        Base64.getEncoder().encodeToString(creds.getBytes(Charset.defaultCharset()));
+    return request.putHeader(AUTHORIZATION.toString(), "Basic " + base64creds);
+  }
+
+  private static void handleRequestException(final Throwable t, final CompletableFuture<?> cf) {
+    cf.completeExceptionally(t);
+  }
+
+  private static <T> void handleResponse(
+      final HttpClientResponse response,
+      final CompletableFuture<T> cf,
+      final ResponseHandlerSupplier<T> responseHandlerSupplier) {
+    if (response.statusCode() == OK.code()) {
+      final RecordParser recordParser = RecordParser.newDelimited("\n", response);
+      final QueryResponseHandler<T> responseHandler =
+          responseHandlerSupplier.get(Vertx.currentContext(), recordParser, cf);
+
+      recordParser.handler(responseHandler::handleBodyBuffer);
+      recordParser.endHandler(responseHandler::handleBodyEnd);
```

Comment:
I think you're right, but I'm unsure what the exception handler should do.

Should we fail the CompletableFuture if it hasn't already been ended? This would require synchronizing the response handler methods, which doesn't seem great.

What if the CompletableFuture has already been completed, as could be the case for streamed queries? Should we close the QueryResult?

Reply:
Ah sorry, forget what I said. If cf is not ended then fail it. But if it's already ended I think the best think is to pass the exception to the response handler (e.g. add a method ResponseHandler.handleException). In the concrete implementation of that method (e.g. in PushQueryResponseHandler.handleException) call sendError on the QueryResultImpl (will need to expose this as it's protected), this will result in the subscriber getting the exception if it's subscribed.

Reply:
Done, though I don't love the method name `QueryResponseHandler#handleExceptionAfterFutureCompleted()`. Let me know if you've got something better haha.

Also, now that I better understand the vertx threading model I realize my concern above is a non-issue: the response handler methods don't need to be synchronized even though the completable future is potentially ended in multiple methods since the different methods are guaranteed to be called from the same thread.